src/core/classes/language-registry.ts

```diff
@@ -0,0 +1,114 @@
+import Language from './language';
+import ComponentRegistry from './registry';
+import type { LanguageProto, Languages } from './language';
+
+export { type ComponentProtoBase } from './registry';
+
+export default class LanguageRegistry extends ComponentRegistry<LanguageProto> {
+	static type: string = 'language';
+	aliases: Record<string, string> = {};
+	instances: Languages = {};
+	defs = new WeakMap<LanguageProto, Language>();
+
+	/**
+	 * Add a language definition to the registry.
+	 * This does not necessarily resolve the language.
+	 */
+	add (def: LanguageProto): boolean {
+		let added = super.add(def);
+
+		if (added && def.alias) {
+			let id = def.id;
+
+			if (typeof def.alias === 'string') {
+				this.aliases[def.alias] = id;
+			}
+			else if (Array.isArray(def.alias)) {
+				for (let alias of def.alias) {
+					this.aliases[alias] = id;
+				}
+			}
+		}
+
+		return added;
+	}
+
+	resolveRef (ref: string | LanguageProto | Language): {
+		id: string;
+		def: LanguageProto;
+		language?: Language;
+	} {
+		if (ref instanceof Language) {
+			return { id: ref.id, def: ref.def, language: ref };
+		}
+
+		let id: string;
+		let def: LanguageProto;
+
+		if (typeof ref === 'object') {
+			def = ref;
+			id = def.id;
+		}
+		else if (typeof ref === 'string') {
+			id = ref;
+		}
+		else {
+			// eslint-disable-next-line @typescript-eslint/restrict-plus-operands
+			throw new Error('Invalid argument type: ' + ref);
+		}
+
+		id = this.aliases[id] ?? id;
+		def ??= this.cache[id];
+
+		return { id, def };
```

Comment:
```suggestion
		let language = this.instances[id];

		return { id, def, language };
```

If there's already a resolved language corresponding to `id`, we should probably return it as well.